typing/env.ml

```diff
@@ -1008,47 +1025,125 @@ let add_required_global id =
   && not (List.exists (Ident.same id) !required_globals)
   then required_globals := id :: !required_globals
 
-let rec normalize_path lax env path =
+let rec normalize_path ~only_absent ~lax env path =
   let path =
     match path with
-      Pdot(p, s, pos) ->
-        Pdot(normalize_path lax env p, s, pos)
+      Pdot(p, s) ->
+        let p = normalize_path ~only_absent ~lax env p in
+        Pdot(p, s)
     | Papply(p1, p2) ->
-        Papply(normalize_path lax env p1, normalize_path true env p2)
+        let p1 = normalize_path ~only_absent ~lax env p1 in
+        let p2 = normalize_path ~only_absent ~lax:true env p2 in
+        Papply(p1, p2)
     | _ -> path
   in
-  try match find_module ~alias:true path env with
-    {md_type=Mty_alias(_, path1)} ->
-      let path' = normalize_path lax env path1 in
-      if lax || !Clflags.transparent_modules then path' else
-      let id = Path.head path in
-      if Ident.global id && not (Ident.same id (Path.head path'))
-      then add_required_global id;
-      path'
+  try match find_module_with_addr ~alias:true path env with
+  | {md_type=Mty_alias(path1)}, addr -> begin
+      match addr with
+      | Some _ when only_absent -> path
+      | _ ->
+          let path' = normalize_path ~only_absent ~lax env path1 in
+          if lax || !Clflags.transparent_modules then path' else
+          let id = Path.head path in
+          if Ident.global id && not (Ident.same id (Path.head path'))
+          then add_required_global id;
+          path'
+    end
```

Comment:
When normalizing, you consider modules that are aliases but have an address. I failed to find where you introduce aliases that have an address different than `None`. Where does that happen?

Reply:
So if a module is present (i.e. it's presence is `Mta_present` in a module type) then it will be given an address (see `store_module` and `components_of_module_maker`). This can happen for aliases than are produced via strengthening (see `strengthen_decl` which can produce aliases, and `strengthen_sig` which then attaches these aliases to a presence that can be `Mta_present`).